### DISCUSSION TITLE
128x32 with 2 Matrices

### DISCUSSION THREAD

**Discussion by brthurr** [MLB-LED-Scoreboard/mlb-led-scoreboard#441] [Q&A] [answered]
I have a RPi3+ using the Adafruit HAT. I have chained two 64x32 matrices together. When I run the following, I get an error:

`sudo python3 main.py --led-gpio-mapping="adafruit-hat" --led-cols=128 --led-rows=64 --led-chain=2`

Error:

```
Suggestion: to slightly improve display update, add
        isolcpus=3
at the end of /boot/cmdline.txt and reboot (see README.md)
ERROR (13:34:29): Invalid matrix dimensions provided. See top of README for supported dimensions.
If you would like to see new dimensions supported, please file an issue on GitHub!
ERROR (13:34:29): Untrapped error in main!
Traceback (most recent call last):
  File "/home/sbadmin/mlb-led-scoreboard/main.py", line 138, in <module>
    main(matrix, config)
  File "/home/sbadmin/mlb-led-scoreboard/main.py", line 29, in main
    config = Config(config_base, matrix.width, matrix.height)
  File "/home/sbadmin/mlb-led-scoreboard/data/config/__init__.py", line 81, in __init__
    json = self.__get_layout(width, height)
  File "/home/sbadmin/mlb-led-scoreboard/data/config/__init__.py", line 228, in __get_layout
    sys.exit(1)
SystemExit: 1
```

I have created a w128h64.json file with the default settings in the example. I hope this is something simple that I am just overlooking.

Any help would be appreciated.

**Comment by pjockey** [comment]
You can add that option to the /boot/config.txt as suggested by the error and It won't hurt you. 
`sudo nano /boot/config.txt` scroll to the end and add the isolcpus=3. Then select ctrl+x and then "y". reboot the Pi.

There was another person in the discussions that had 2 displays connected together and he was using this command
`sudo python3 main.py --led-rows=32 --led-cols=64 --led-chain=4 --led-gpio-mapping=adafruit-hat-pwm --led-pixel-mapper=U-mapper --led-brightness=30`

I don't think you need the "" around adafruit-hat.

Do you have the anti-flicker fix done on your Hat? That is the pwm option at the end of the adafruit-hat.

**Comment by ty-porter** [comment]
Two 64x32 matrices chained are either 128x32 or 64x64, not 128x64. If you were intending 128x64 then you will need to chain two additional 64x32 matrices in a U pattern.

Secondly when you are overriding the `.example` coordinates, make sure not to remove those files (copy them to a new file such as `w128h64.json` instead). The `.example` reference files are used to determine which matrix dimensions are supported.

**Reply by brthurr** [reply]
I didn't remove it. Just made a copy and didn't change anything.

I removed the incorrect `w128h64.json` and used `w128h32.json` instead. It runs but the second matrix is not displaying properly:

![20230331_194016136_iOS conv](https://user-images.githubusercontent.com/11654145/229215247-0d07f8b8-37ed-4849-b9cc-687e039c6f30.jpeg)

When I use `--led-gpio-mapping=adafruit-hat-pwm` instead of `--led-gpio-mapping=adafruit-hat-pwm`, I get this:

`=== snd_bcm2835: found that the Pi sound module is loaded. ===`

Which is bogus because I have disabled the driver in /boot/config.txt

```
❯ cat /boot/config.txt | grep audio
# uncomment to force a HDMI mode rather than DVI. This can make audio work in
# Enable audio (loads snd_bcm2835)
# dtparam=audio=on
```

**Reply by ty-porter** [reply, marked as answer]
This looks like an issue with your hardware. I would recommend breaking the problem down.

I suggest trying to run a single 64x32 matrix first against the default coordinates file. You shouldn't have to copy from `w64h32.json.example` to `w64h32.json`, it will handle that automatically.

The error you're describing regarding `snd_bcm2835` is for the PWM anti-flicker mod, I would recommend ignoring that GPIO mapping option until you can diagnose what's going on with the matrix.

One thing that might be an issue for daisy chains of this size is that single matrices >= 64px in height requires soldering the E address pin on the bonnet: https://learn.adafruit.com/adafruit-rgb-matrix-plus-real-time-clock-hat-for-raspberry-pi/assembly#64x64-matrices-solder-e-jumper-3004286 (I'm unable to confirm this as my hardware does not match yours).

Some additional resources:
[CLI flags](https://github.com/MLB-LED-Scoreboard/mlb-led-scoreboard#flags)
[Driver library](https://github.com/hzeller/rpi-rgb-led-matrix)

**Reply by brthurr** [reply]
I have tested with both matrices individually. They work fine. I'll solder the E address now and see if that does it. I'll deal with the PWM issue later.

I'll let you know.

**Reply by brthurr** [reply]
That did it! As for the `snd_bcm2835` error, looks like there's additional soldering that has to be performed on the HAT. That's a tomorrow project. Today, I get to play. Thanks for all of your help!

**Reply by mdskier** [reply]
Are you using the 2 32x64 displays side by side or one on top of the other?
